base/learner.h

```diff
@@ -0,0 +1,45 @@
+#ifndef  _MACHINELEARNING_BASE_LEARNER_H_ 
+#define  _MACHINELEARNING_BASE_LEARNER_H_
+
+#include <vector>
+
+namespace base {
+
+/**
+ * Base class for all learning algorithm implementations.
+ */
+class Learner{
+
+  public:
+
+    /**
+     * \brief Standard destructor.
+     */
+    virtual ~Learner();
+
+    /**
+     * Train the learner with a given dataset.
+     * \param X A pointer object of Collection.
+     */
+    virtual void Train(/*const Collection *X*/);
+
+    /**
+     * Predicts the labels of the instances of X based on the fitted model.
+     * \param X           A pointer object of Collection.
+     * \param prediction  Instances prediction.
+     */
+    virtual void Predict(/*const Collection *X,*/ std::vector<double> &prediction) = 0;
```

Comment:
Returning a vector is by far the worst idea. At some point during execution you will have 2 copies of the data floating around memory. When handling big collections that's prohibitive. The best way to handle this is to pass the vector where you want to store the data as a pointer. So the answer is: